### PR TITLE
etcdutil, mcs: fix the issue loading label rules is too slow

### LINKED ISSUE #7724: mcs: loading label rules is too slow
### What did you do?

test mcs in dev env with 20000+ label

### What did you expect to see?

run normally

### What did you see instead?

It loaded for more than 10 minutes without successful startup

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #7724

### What is changed and how does it work?

```commit-message
```
1. cancel timeout in loading function
2. use larger batch when loading from etcd
3. reduce times that call `buildRangeList` in label watcher

### Check List

Tests

- Unit test
```
goos: linux
goarch: amd64
pkg: github.com/tikv/pd/pkg/mcs/scheduling/server/rule
cpu: AMD Ryzen 7 7840S with Radeon 780M Graphics         
```
| BenchmarkLoadLargeRules-16      |  | ||
| :---        |    ----: |        ----:  | ----: | 
| Master(timeout1h)     |  270616573162 ns/op        | 199973994432 B/op     |270541132 allocs/op|
| batch | 303262574 ns/op|	135643794 B/op	| 1146309 allocs/op|
| batch with larger limit (This PR) |208206918 ns/op |       128028446 B/op       |  1197805 allocs/op|
| This PR with easyjson | 201611560 ns/op       | 124696554 B/op    |     1132302 allocs/op|

#### cpu time

Master:
image

This PR:
image

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/utils/etcdutil/etcdutil.go [hunk 1]
```diff
@@ -902,6 +893,17 @@ func (lw *LoopWatcher) load(ctx context.Context) (nextRevision int64, err error)
 		if err != nil {
 			log.Error("load failed in watch loop", zap.String("name", lw.name),
 				zap.String("key", lw.key), zap.Error(err))
+			if strings.Contains(err.Error(), codes.ResourceExhausted.String()) {
```
Comment (rleungx) [inline pkg/utils/etcdutil/etcdutil.go:896]: Do we need this check?

File: pkg/utils/etcdutil/etcdutil.go [hunk 2]
```diff
@@ -902,6 +893,17 @@ func (lw *LoopWatcher) load(ctx context.Context) (nextRevision int64, err error)
...
+				if limit == 0 {
+					limit = maxLoadBatchSize
+				} else if limit > minLoadBatchSize {
+					limit /= 2
+				} else {
+					return 0, err
+				}
+				opts = append(opts, clientv3.WithLimit(limit+1))
```
Comment (rleungx) [inline pkg/utils/etcdutil/etcdutil.go:904]: Will the option become more and more?

File: pkg/schedule/labeler/labeler.go [hunk 3]
```diff
@@ -128,11 +128,12 @@ func (l *RegionLabeler) loadRules() error {
...
 		}
 	}
-	l.buildRangeList()
+	l.BuildRangeList()
 	return nil
 }
 
-func (l *RegionLabeler) buildRangeList() {
+// BuildRangeList builds the range list.
+func (l *RegionLabeler) BuildRangeList() {
```
Comment (rleungx) [inline pkg/schedule/labeler/labeler.go:136]: BuildRangeListLocked

File: pkg/utils/etcdutil/etcdutil.go (hunk 1 above)
Comment (lhy1024) [reply pkg/utils/etcdutil/etcdutil.go]: add more check

File: pkg/utils/etcdutil/etcdutil.go [hunk 4]
```diff
@@ -896,12 +887,29 @@ func (lw *LoopWatcher) load(ctx context.Context) (nextRevision int64, err error)
...
+		default:
+		}
 		// Sort by key to get the next key and we don't need to worry about the performance,
 		// Because the default sort is just SortByKey and SortAscend
-		resp, err := clientv3.NewKV(lw.client).Get(ctx, startKey, opts...)
+		resp, err := EtcdKVGet(lw.client, startKey, opts...)
 		if err != nil {
 			log.Error("load failed in watch loop", zap.String("name", lw.name),
 				zap.String("key", lw.key), zap.Error(err))
+			if strings.Contains(err.Error(), codes.ResourceExhausted.String()) ||
```
Comment (rleungx) [inline pkg/utils/etcdutil/etcdutil.go:901]: Only two errors?

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 5]
```diff
@@ -211,27 +215,42 @@ func (rw *Watcher) initializeRuleWatcher() error {
 }
 
 func (rw *Watcher) initializeRegionLabelWatcher() error {
+	if rw.regionLabeler == nil {
```
Comment (rleungx) [inline pkg/mcs/scheduling/server/rule/watcher.go:218]: When will it happen?

**Review by rleungx — APPROVED** [review]
The rest LGTM

**Review by CabinfeverB — APPROVED** [review]

**Pull request merged**